### DISCUSSION TITLE
Why are ParamSpecs always invariant, rather than contravariant?

### DISCUSSION THREAD

**Discussion by EthanBullVulpe** [microsoft/pyright#5751] [Q&A]
Hello, I am wondering why pyright treats all `ParamSpec`s as invariant.

There is some discussion of `ParamSpec`s and variance here: https://github.com/python/typing/issues/1027
It seems like the conclusion there was that the idea of variance was not well-defined or simple for `ParamSpec`s, and so to treat all `ParamSpec`s as invariant for now.

However, treating them as always invariant, leads to some limitations, for example:
```
from __future__ import annotations
from abc import abstractmethod

from typing import Callable, Generic
from typing_extensions import TypeVar, ParamSpec

Input = ParamSpec("Input")
Output = TypeVar("Output", covariant=True)

class FancyCallable(Generic[Input, Output]):
    
    @abstractmethod
    def __call__(self, *args: Input.args, **kwargs: Input.kwargs) -> Output:
        raise NotImplementedError

class Super:
    ...
    
class Sub(Super):
    ...
    
fancy1: FancyCallable[[Super], int] = FancyCallable[[Super], int]()
fancy2: FancyCallable[[Sub], int] = fancy1 # This assignment should be allowed, but since the ParamSpec `Input` is invariant, a reportGeneralTypingIssues is reported.

callable1: Callable[[Super], int] = FancyCallable[[Super], int]()
callable2: Callable[[Sub], int] = callable1 # If ordinary Callables are used, the assignment is allowed, because the Callable ParamSpec seems to be contravariant.
```

In this snippet the assignment `fancy2: FancyCallable[[Sub], int] = fancy1` should really be allowed, since any `FancyCallable` that can take in a `Super` can also take in a `Sub`. However, pyright does not allow this assignment, since the `Input` `ParamSpec` is invariant.

The similar assignment using `Callable`s instead of `FancyCallable`s is accepted, so clearly pyright already has some understanding of `ParamSpec` contravariance.

What I am wondering is:
Since `ParamSpec`s can only be used as function inputs (through `ParamSpec.args` and `ParamSpec.kwargs` annotations), why not always treat them as contravariant instead of invariant?

Is there a possible case where it would be incorrect to treat a `ParamSpec` as contravariant?

**Comment by erictraut** [comment]
I think (but could be convinced otherwise with a compelling argument) that the notion of "variance" makes no sense when talking about ParamSpec. I can understand why you might think that ParamSpecs are contravariant because the types of the individual parameters that make up a callable's signature are generally contravariant. But when you consider the entire signature, things are more complicated.

Consider the following pairs of signatures, which have subtype relationships. Note that I haven't varied the parameter types in any of these pairs — they're all `int`.
1. `(a: int)` and `(a: int, /)`
2. `(a: int)` and `(a: int, b: int = ...)`
3. `(*args: int)` and `(a: int, b: int)`

I said "generally contravariant" above because this isn't always the case. Consider the following modification to your sample above:

```python
class FancyCallable2[**Input, Output]:
    def __call__(self, x: Callable[Input, Output]) -> Output:
        raise NotImplementedError
```

Here the ParamSpec `Input` is used for a callback, so any variance rules effectively flip. And if we nest another callback, it flips again, as in the following:

```python
class FancyCallable3[**Input, Output]:
    def __call__(self, x: Callable[[Callable[Input, Output], Output], Output]) -> Output:
        raise NotImplementedError
```

**Reply by EthanBullVulpe** [reply]
I see what you mean about the variance of the ParamSpec reversing each time it is included in a once-deeper nested callback. So clearly, my initial thought that all ParamSpecs could be treated as contravariant is incorrect.

However, I do think the idea of "variance" is well-defined for ParamSpecs, though I am interested in your thoughts.

As you point out in your comment, signatures (and thus ParamSpecs) do have subtype relationships.
1. `(a: int)` is a subtype of `(a: int, /)`
2. `(a: int)` is (i believe) a supertype of `(a: int, b: int = ...)`
3. `(*args: int)` is a subtype of  `(a: int, b: int)`

To the best of my understanding, what defines the subtype relationships between signatures is "signature A is a subtype of signature B if any valid way to call B is also a valid way to call A".

Then, based on this definition of subtype relationships between signatures, the definitions of contravariance, covariance, and invariance follow in the same way as they do for ordinary types which have subtype relationships.

Do you think this is correct?

**Reply by erictraut** [reply]
After giving this more thought, I think I agree with your perspective.

I think we can say in general that `Callable[P1, R1]` is a subtype of `Callable[P2, R2]` if `P1` is a subtype of `P2` _and_ `R1` is a subtype of `R2`. That means a ParamSpec within a `Callable` is covariant. Of course, the individual parameters that make up a param spec are contravariant in their types. 

I think this underscores the need for someone to formally write down the subtyping rules for callables and param specs. This has come up a number of times in recent discussions (including several bug reports in pyright and mypy and the ongoing talks about intersection types). I'll try to write some documentation when I have time.
